### ISSUE TITLE
Feature request: Add title attribute for iframes created from Wagtail embeds

### ISSUE THREAD

**Issue by thibaudcolas** [wagtail/wagtail#5982]
### Is your proposal related to a problem?

Wagtail creates iframes without `title` with its EmbedBlock and embed chooser / rich text `embed` feature. iframes without `title` are just announced as "frame" for screen reader users, without any further details, which isn’t enough for them to determine what the frame contains, and decide whether to enter it or skip it.

- Accessibility guidelines on `title` for iframes: https://www.w3.org/TR/WCAG20-TECHS/H64.html
- Example of the problem, with rich text embed and StreamField EmbedBlock: https://bakerydemo-thibaudcolas.herokuapp.com/breads/one-more-bread/
- Example on a real-world site (that may change): https://www.buckinghamshire.gov.uk/news/teacher-and-students-pay-musical-tribute-uks-key-workers/
- Demo video: https://youtu.be/eRKcWXjw61Y

Note: I believe this is also a problem with Wagtail’s `embed` template tag, which [doesn’t seem to support setting a title](https://docs.wagtail.io/en/latest/advanced_topics/embeds.html).

#### Steps to Reproduce

1. Create an embed, either with the EmbedBlock or rich text "Embed"
2. Save / publish the page
3. Access the page with a screen reader
4. Try to navigate the page sequentially, or with region navigation (rotor for VoiceOver)

#### Technical details

* Browser version: Latest Chrome and Safari with VoiceOver on macOS 10.14.

### Describe the solution you'd like

I can see two potential fixes for this,

- Either we add a "Title" field to both the embed chooser and the EmbedBlock, with relevant help text for editors (e.g. "Concise description of the embed for screen reader users – for example, title of the embedded content")
- Or we automatically populate the title based on the embed’s metadata. I believe embeds already have a title field, but I’m not sure whether it’s always filled, and if so with what kind of content

### Describe alternatives you've considered

An alternative would be to create a custom Embed block / rich text feature that is more easily customisable / more screen-reader-friendly, however I believe this is better suited as an enhancement directly in Wagtail

### Additional context

This might be further complicated by the fact that we might just be inserting the oembed / Embedly provided HTML into the template, rather than something crafted by Wagtail – in which case it would be good to see whether Wagtail can add a `title` in there, or otherwise we could:

- Provide an opt-in way to use Wagtail’s own `<iframe>` generation instead of the platform-specific `html`. 
- At least document the fact that this is an issue.

**Comment by thibaudcolas** [comment]
cc @zmustapha and @rjpruitt16 who made similar improvements in #5235 and #5446, and might have some thoughts on how to improve this.

**Comment by thibaudcolas** [comment]
It seems that at least Vimeo gets this right,

```html
<iframe src="https://player.vimeo.com/video/412325301?app_id=122963" width="426" height="178" frameborder="0" allow="autoplay; fullscreen" allowfullscreen="" title="Fortem People - Our Third CRM System (Trailer)"></iframe>
```

So far I’ve seen this issue with YouTube and SoundCloud embeds.

**Comment by thibaudcolas** [comment]
Here is an approach to solving this from CodeRed: https://github.com/coderedcorp/coderedcms/pull/207

```py
@register.simple_tag
def render_iframe_from_embed(embed):
    soup = BeautifulSoup(embed.html, "html.parser")
    try:
        iframe_tags = soup.find('iframe')
        iframe_tags['title'] = embed.title
        return mark_safe(soup.prettify())
    except AttributeError:
        pass
    except TypeError:
        pass

    return mark_safe(embed.html)
```

```html
{% render_iframe_from_embed embed %}
```

It feels a bit wasteful to me to do this when rendering templates rather than when saving embeds, but has the advantage of not needing migrations.

**Comment by thibaudcolas** [comment]
Related to the above – pages with YouTube embeds from Wagtail also fail HTML validation with https://validator.github.io/validator/ due to the `frameborder` attribute:

```
<page>":832.5-832.212: error: The “frameborder” attribute on the “iframe” element is obsolete. Use CSS instead.
```

If we fix the `title` issue by changing the HTML of the embed, it would be great to either fix `frameborder` as well by default, or have a way to opt into that additional change.

**Comment by thibaudcolas** [comment]
Tagging @nimasmi who has recently fixed this on a project in a way I find quite elegant. We may need to validate with other core developers that they are happy with his approach being backported to Wagtail, if so this could be quite a nice fix.

**Comment by thibaudcolas** [comment]
As of a few days ago, Wagtail core now has a few custom embed finders for other specific providers: https://github.com/wagtail/wagtail/pull/6550. So it feels like @nimasmi’s implementation would also be a natural fit, assuming Wagtail is ok to have more of such provider-specific finders.

I’ll tag this as "needs design decision" because we to agree on:

- Whether a custom finder is indeed the most appropriate way to do this, or whether a Wagtail-core fix would be better done in the base OEmbed finder, or keep finders as-is and only change the embeds’ rendering
- If using a custom finder, we need to decide for which providers to set it up. YouTube is where there’s an obvious need for this, but I’d also like to review a few other common embeds to see how they fare.

@nimasmi do you have thoughts on what might be an appropriate solution for Wagtail core?

**Comment by thibaudcolas** [comment]
Here is @nimasmi’s implementation for easier reference, with any project-specific customizations removed. I don’t know why this only happens for `video` embeds, I think I’d be keen to do this for all embeds although I’ve only had reports of this issue for YouTube so far.

Settings:

```python
# settings.py
from mysite.utils.embed_finders import CustomOEmbedFinder

WAGTAILEMBEDS_FINDERS = [
    {"class": "mysite.utils.embed_finders.CustomOEmbedFinder"},
]
```

Custom finder:

```python
# embed_finders.py
from wagtail.embeds.finders.oembed import OEmbedFinder
from wagtail.embeds.oembed_providers import youtube

from bs4 import BeautifulSoup

class CustomOEmbedFinder(OEmbedFinder):
    """OEmbed finder to set video iframe titles."""

    def find_embed(self, url, max_width=None):
        embed = super().find_embed(url, max_width)
        if embed["type"] == "video":
            soup = BeautifulSoup(embed["html"], "html.parser")
            iframe = soup.find("iframe")

            # a11y: set iframe title
            try:
                iframe.attrs["title"] = embed["title"]
            except KeyError:
                pass

            embed["html"] = str(soup)

        return embed
```

Test cases:

```python
# test_embed_finders.py
from unittest.mock import patch

from django.test import TestCase

from wagtail.embeds import oembed_providers
from wagtail.embeds.finders import get_finders

from bs4 import BeautifulSoup

from mysite.utils.embed_finders import CustomOEmbedFinder

@patch("mysite.utils.embed_finders.OEmbedFinder.find_embed")
class CustomOEmbedFinderTest(TestCase):
    def test_video_iframe_title_is_set(self, mock_find_embed):
        mock_find_embed.return_value = {
            "html": '<iframe src="www.example.com"></iframe>',
            "type": "video",
            "title": "Some video",
        }

        finder = CustomOEmbedFinder()
        result = finder.find_embed("www.example.com")
        soup = BeautifulSoup(result["html"], "html.parser")
        self.assertEqual(soup.find("iframe").attrs.get("title"), "Some video")

    def test_nonvideo_iframe_title_is_not_set(self, mock_find_embed):
        mock_find_embed.return_value = {
            "html": '<iframe src="www.example.com"></iframe>',
            "type": "not a video",
            "title": "Not a video",
        }

        finder = CustomOEmbedFinder()
        result = finder.find_embed("www.example.com")
        soup = BeautifulSoup(result["html"], "html.parser")
        self.assertEqual(soup.find("iframe").attrs.get("title"), None)

    def test_video_output_html(self, mock_find_embed):
        mock_find_embed.return_value = {
            "html": '<iframe src="www.example.com"></iframe>',
            "type": "video",
            "title": "Some video",
        }

        finder = CustomOEmbedFinder()
        result = finder.find_embed("www.example.com")
        self.assertEqual(
            result["html"],
            '<iframe src="www.example.com" title="Some video"></iframe>',
        )

    def test_nonvideo_output_html(self, mock_find_embed):
        mock_find_embed.return_value = {
            "html": '<iframe src="www.example.com"></iframe>',
            "type": "not a video",
            "title": "Something else",
        }

        finder = CustomOEmbedFinder()
        result = finder.find_embed("www.example.com")
        self.assertEqual(
            result["html"], '<iframe src="www.example.com"></iframe>',
        )
```

**Comment by thibaudcolas** [comment]
It looks like YouTube has (finally!) fixed this on their side – so I’ll close this now.

**Comment by nimasmi** [comment]
:relieved:
